File: Frontend/STUDER/src/app/features/contest/contest.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { API_CONFIG } from '../../config/api.config';
import {
  ContestResponseDTO, ContestCreateRequestDTO, ContestCourseResponseDTO,
  CourseRatingRequestDTO, LeaderboardEntryDTO, AchievementResponseDTO,
  DashboardMetricsDTO, StatusChangeRequest
} from './contest.model';
import { CourseResponseDTO, CourseCreateRequestDTO } from '../courses/course.model';
import { MessageResponseDTO } from '../discussions/discussion.model';

@Injectable({ providedIn: 'root' })
export class ContestService {
  private readonly base = `${API_CONFIG.baseUrl}${API_CONFIG.contests}`;
  private readonly adminBase = `${API_CONFIG.baseUrl}${API_CONFIG.admin}`;

  constructor(private readonly http: HttpClient) {}

  // Public
  list(status?: string, page = 0, size = 10): Observable<any> {
    let params = `?page=${page}&size=${size}`;
    if (status) params += `&status=${status}`;
    return this.http.get<any>(`${this.base}${params}`);
  }

  getById(id: number): Observable<ContestResponseDTO> {
    return this.http.get<ContestResponseDTO>(`${this.base}/${id}`);
  }

  getActive(): Observable<ContestResponseDTO[]> {
    return this.http.get<ContestResponseDTO[]>(`${this.base}/active`);
  }

  getUpcoming(): Observable<ContestResponseDTO[]> {
    return this.http.get<ContestResponseDTO[]>(`${this.base}/upcoming`);
  }

  submitCourse(id: number, data: CourseCreateRequestDTO): Observable<CourseResponseDTO> {
    return this.http.post<CourseResponseDTO>(`${this.base}/${id}/submit`, data);
  }

  getRandomValidationCourse(id: number): Observable<ContestCourseResponseDTO> {
    return this.http.get<ContestCourseResponseDTO>(`${this.base}/${id}/validate/random`);
  }

  rateCourse(data: CourseRatingRequestDTO): Observable<MessageResponseDTO> {
    return this.http.post<MessageResponseDTO>(`${this.base}/validate`, data);
  }

  getLeaderboard(id: number): Observable<LeaderboardEntryDTO[]> {
    return this.http.get<LeaderboardEntryDTO[]>(`${this.base}/${id}/leaderboard`);
  }

  getAchievements(): Observable<AchievementResponseDTO[]> {
    return this.http.get<AchievementResponseDTO[]>(`${this.base}/achievements`);
  }

  getAllAchievements(): Observable<AchievementResponseDTO[]> {
    return this.http.get<AchievementResponseDTO[]>(`${this.base}/achievements/all`);
  }

  // Admin
  getDashboard(): Observable<DashboardMetricsDTO> {
    return this.http.get<DashboardMetricsDTO>(`${this.adminBase}/dashboard`);
  }

  createContest(data: ContestCreateRequestDTO): Observable<ContestResponseDTO> {
    return this.http.post<ContestResponseDTO>(`${this.adminBase}/contests`, data);
  }

  updateContest(id: number, data: ContestCreateRequestDTO): Observable<ContestResponseDTO> {
    return this.http.put<ContestResponseDTO>(`${this.adminBase}/contests/${id}`, data);
  }

  deleteContest(id: number): Observable<MessageResponseDTO> {
    return this.http.delete<MessageResponseDTO>(`${this.adminBase}/contests/${id}`);
  }

  changeStatus(id: number, status: string): Observable<MessageResponseDTO> {
    return this.http.patch<MessageResponseDTO>(`${this.adminBase}/contests/${id}/status`, { status });
  }

  finishContest(id: number): Observable<MessageResponseDTO> {
    return this.http.post<MessageResponseDTO>(`${this.adminBase}/contests/${id}/finish`, {});
  }
}

File: Backend/STUDER/src/main/java/facu/studer/services/ContestService.java
package facu.studer.services;

import facu.studer.DTOs.MessageDTO;
import facu.studer.DTOs.contest.*;
import facu.studer.DTOs.courses.CourseCreateRequestDTO;
import facu.studer.DTOs.courses.CourseResponseDTO;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.List;

public interface ContestService {
    // Public
    ContestResponseDTO getContest(Long contestId);
    Page<ContestResponseDTO> listContests(String status, Pageable pageable);
    Page<ContestResponseDTO> searchContests(String query, Pageable pageable);
    List<ContestResponseDTO> getActiveContests();
    List<ContestResponseDTO> getUpcomingContests();

    // Participation
    CourseResponseDTO submitCourseForContest(String username, Long contestId, CourseCreateRequestDTO request);
    ContestCourseResponseDTO getRandomCourseForValidation(String username, Long contestId);
    MessageDTO rateCourse(String username, CourseRatingRequestDTO request);
    List<LeaderboardEntryDTO> getLeaderboard(Long contestId);

    // Admin
    ContestResponseDTO createContest(String username, ContestCreateRequestDTO request);
    ContestResponseDTO updateContest(String username, Long contestId, ContestCreateRequestDTO request);
    MessageDTO deleteContest(String username, Long contestId);
    MessageDTO changeContestStatus(String username, Long contestId, String newStatus);
    MessageDTO finishContest(String username, Long contestId);

    // Achievements
    List<AchievementResponseDTO> getUserAchievements(String username);
    List<AchievementResponseDTO> getAllAchievements();
}
